### PR TITLE
feat: implement frobenius map

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This implement https://en.wikipedia.org/wiki/Frobenius_endomorphism

### CODE REVIEW

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

File: tachyon/math/finite_fields/fp3.h [hunk 1]
```diff
@@ -27,7 +29,65 @@ class Fp3 final : public CubicExtensionField<Fp3<Config>> {
...
+
+    // kFrobeniusCoeffs[1] = q^(exp₁) = q^((P¹ - 1) / 3) = ω
+    SET_FROBENIUS_COEFF(1);
+    // kFrobeniusCoeffs[2] = q^(exp₂) = q^((P² - 1) / 3)
+    SET_FROBENIUS_COEFF(2);
+
+#undef SET_FROBENIUS_COEFF
+
+    // kFrobeniusCoeffs[0] = q^(2 * (P⁰ - 1) / 3) = 1
+    Config::kFrobeniusCoeffs2[0] = FrobeniusCoefficient::One();
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/finite_fields/fp3.h:79]: For cubic extension field, it is represented in `c0 + c1x + c2x^2`. If you powers to `p`, then it is evaluated to `c0 + c1^p * x^p + c2^p * x^2p`. If it is `fp3`, then `c0` remains untouched, `c1` should be multiplied by `kFrobeniusCoeff[1]` and `c2` should be multiplied by `kFrobeniusCoeff2[1]`

File: tachyon/math/finite_fields/cubic_extension_field.h [hunk 2]
```diff
@@ -60,6 +60,40 @@ class CubicExtensionField : public CyclotomicMultiplicativeSubgroup<Derived> {
...
+    // never overflows even on 32 bit machine.
+    size_t index_multiplier = size_t{BaseField::ExtensionDegree()};
+    CubicExtensionField self_to_p = *this;
+    self_to_p.FrobeniusMapInPlace(index_multiplier);
+    CubicExtensionField self_to_p2 = *this;
+    self_to_p2.FrobeniusMapInPlace(2 * index_multiplier);
+    self_to_p *= (self_to_p2 * (*this));
+    // NOTE(chokobole): below CHECK() is not a device code.
+    // See https://github.com/kroma-network/tachyon/issues/76
+    // CHECK(!(self_to_p.c1().IsZero() && self_to_p.c2().IsZero()));
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/finite_fields/cubic_extension_field.h:82]: Should be uncommented!

File: tachyon/math/finite_fields/fp3.h [hunk 3]
```diff
@@ -27,7 +29,65 @@ class Fp3 final : public CubicExtensionField<Fp3<Config>> {
...
+#define SET_FROBENIUS_COEFF2(d)                              \
+  gmp::CopyLimbs(mpz_class(2) * exp##d##_gmp, exp##d.limbs); \
+  Config::kFrobeniusCoeffs2[d] = Config::kNonResidue.Pow(exp##d)
+
+    // kFrobeniusCoeffs[1] = q^(2 * exp₁) = q^(2 * (P¹ - 1) / 3) = ω²
+    SET_FROBENIUS_COEFF2(1);
+    // kFrobeniusCoeffs[2] = q^(2 * exp₂) = q^(2 * (P² - 1) / 3)
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/fp3.h:86]:
```suggestion
    // kFrobeniusCoeffs2[0] = q^(2 * (P⁰ - 1) / 3) = 1
    Config::kFrobeniusCoeffs2[0] = FrobeniusCoefficient::One();
#define SET_FROBENIUS_COEFF2(d)                              \
  gmp::CopyLimbs(mpz_class(2) * exp##d##_gmp, exp##d.limbs); \
  Config::kFrobeniusCoeffs2[d] = Config::kNonResidue.Pow(exp##d)

    // kFrobeniusCoeffs2[1] = q^(2 * exp₁) = q^(2 * (P¹ - 1) / 3) = ω²
    SET_FROBENIUS_COEFF2(1);
    // kFrobeniusCoeffs2[2] = q^(2 * exp₂) = q^(2 * (P² - 1) / 3)
```

File: tachyon/math/finite_fields/fp3.h [hunk 4]
```diff
@@ -27,7 +29,65 @@ class Fp3 final : public CubicExtensionField<Fp3<Config>> {
...
+  Config::kFrobeniusCoeffs[d] = Config::kNonResidue.Pow(exp##d)
+
+    // kFrobeniusCoeffs[1] = q^(exp₁) = q^((P¹ - 1) / 3) = ω
+    SET_FROBENIUS_COEFF(1);
+    // kFrobeniusCoeffs[2] = q^(exp₂) = q^((P² - 1) / 3)
+    SET_FROBENIUS_COEFF(2);
+
+#undef SET_FROBENIUS_COEFF
+
+    // kFrobeniusCoeffs[0] = q^(2 * (P⁰ - 1) / 3) = 1
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/fp3.h:78]:
```suggestion
    // kFrobeniusCoeffs[0] = q^((P⁰ - 1) / 3) = 1
```

Comment (chokobole) [reply tachyon/math/finite_fields/fp3.h:78]: I think this should be

```suggestion
    // kFrobeniusCoeffs2[0] = q^(2 * (P⁰ - 1) / 3) = 1

```

Comment (TomTaehoonKim) [reply tachyon/math/finite_fields/fp3.h]: @chokobole Oh then should https://github.com/kroma-network/tachyon/pull/101/files#diff-e673446d9e3e55b20b8c313a638d62f9e9af8b2135d6dd4745d5e3ca601fadf1R182 be fixed with 2 * at the front?

Comment (chokobole) [reply tachyon/math/finite_fields/fp3.h]: You're right! I'll fix it.

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by Insun35 — APPROVED** [review]
LGTM

**Pull request merged**